mypy/types.py

```diff
@@ -1496,7 +1496,14 @@ def is_singleton_type(self) -> bool:
     def get_enum_values(self) -> list[str]:
         """Return the list of values for an Enum."""
         return [
-            name for name, sym in self.type.names.items() if isinstance(sym.node, mypy.nodes.Var)
+            name
+            for name, sym in self.type.names.items()
+            if (
+                isinstance(sym.node, mypy.nodes.Var)
+                and name not in ENUM_REMOVED_PROPS
+                and not name.startswith("__")
+                and sym.node.has_explicit_value
```

Comment:
This last condition changes the behavior of mypy which currently considers unassigned but annotated attributes as members of the enum.
I would say the new behavior is better but it must be noted in the PR title or description that this is a breaking change that must be considered on its own.
Looking at the mypy primer output, it looks like members of enums constructed using the enum call syntax do not have the `has_explicit_value` set which breaks type narrowing. I suggest changing the variable [here](https://github.com/python/mypy/blob/fb31409b392c5533b25173705d62ed385ee39cfb/mypy/semanal_enum.py#L143-L146) to also set this flag perhaps with a comment explaining the reason -- something like "`E = Enum("E", "A")` is equivalent to `class E(Enum): A = auto()`"